src/compiler/emitter.ts

```diff
@@ -2302,35 +2394,42 @@ namespace ts {
         function emitModifiers(node: Node, modifiers: NodeArray<Modifier>) {
             if (modifiers && modifiers.length) {
                 emitList(node, modifiers, ListFormat.Modifiers);
-                write(" ");
+                writeSpace(" ");
             }
         }
 
-        function emitWithPrefix(prefix: string, node: Node) {
-            emitNodeWithPrefix(prefix, node, emit);
+        function emitWithPrefix(prefix: string | [string, string, string], writer: (s: string) => void, node: Node) {

```

Comment:
Our emit time keeps creeping upward. I'm wary of making any function more polymorphic than necessary. Can you verify the performance characteristics of this change?

Reply:
@rbuckton I can refactor all of these callsites to be non-polymorphic (`emitWithPrefix`/`emitWithMultipartPrefix`) pretty easily. I definitely get your concern (hence why there's only one copy of each prefix/postfix array that the multipart prefix/postfix uses). However, I'm _pretty_ certain this will inevitably cause our emit time to increase some _no matter what_, since the node builder inherently has more overhead than direct string building (it actually allocates ASTs, after all) - so it's a matter of identifying everywhere an improvement can be made to keep the change minimal.

Reply:
I'd go so far as to say that, since there are only 3 triples right now, I'd create dedicated functions for the first two triples (used 10 and 6 times, respectively, something like `emitAsTypeAnnotation`) and inline the last, which is only used once.